dvc/repo/diff.py

```diff
@@ -85,3 +47,63 @@ def _exists(output):
     }
 
     return ret if any(ret.values()) else {}
+
+
+def _paths_checksums(repo):
+    """
+    A dictionary of checksums addressed by relpaths collected from
+    the current tree outputs.
+
+    To help distinguish between a directory and a file output,
+    the former one will come with a trailing slash in the path:
+
+        directory: "data/"
+        file:      "data"
+    """
+
+    return dict(_output_paths(repo))
+
+
+def _output_paths(repo):
+    on_working_tree = isinstance(repo.tree, LocalTree)
```

Comment:
Does `on_working_tree` matter anymore though? `RepoTree` will just use `repo.tree` and will act accordingly if it is a wtree or a git rev tree.

Reply:
So the issue here is that DvcTree (and RepoTree) does not re-compute dir output hashes when the workspace is dirty. I'm not sure that we would want to change that behavior just for `dvc diff`, it seems like in a majority of scenarios we would want to re-use the already computed hashes for dir outputs when we are using DvcTree and RepoTree.

Reply:
That's actually an inconsistency that should be solved in RepoTree. Ideally, we should be able to work with both dirty state and clean repos. But could start with the current approach too.

Reply:
Filed #4523 for this issue